src/deluge/memory/cache_manager.cpp

```diff
@@ -0,0 +1,176 @@
+					AudioEngine::bypassCulling = true;
+				}
+				stealable = static_cast<Stealable*>(reclamation_queue_[q].getNext(stealable));
+				continue;
+			}
+
+			// If we're not in the last queue, and we haven't tried this too many times yet, check whether it was actually in the right queue
+			if (q < NUM_STEALABLE_QUEUES - 1 && numberReassessed < 4) {
+				numberReassessed++;
+
+				int appropriateQueue = stealable->getAppropriateQueue();
+
+				// If it was in the wrong queue, put it in the right queue and start again with the next one in our queue
+				if (appropriateQueue > q) {
+
+					Debug::print("changing queue from ");
+					Debug::print(q);
+					Debug::print(" to ");
+					Debug::println(appropriateQueue);
+
+					auto* next = static_cast<Stealable*>(reclamation_queue_[q].getNext(stealable));
+
+					stealable->remove();
+					this->QueueForReclamation(appropriateQueue, stealable);
+
+					stealable = next;
+					continue;
+				}
+			}
+
+			// Ok, we've got one Stealable
+			uint32_t* __restrict__ header = (uint32_t*)((uint32_t)stealable - 4);
+			spaceSize = (*header & SPACE_SIZE_MASK);
+
+			stealable->lastTraversalNo = currentTraversalNo;
+
+			// How much additional space would we need on top of this Stealable?
+			int amountToExtend = totalSizeNeeded - spaceSize;
+
+			newSpaceAddress = (uint32_t)stealable;
+
+			// If that one Stealable alone was big enough, that's great
+			if (amountToExtend <= 0) {
+				found = true;
+				break;
+			}
+
+			// Otherwise, see if available neighbouring memory adds up to make enough in total
+			NeighbouringMemoryGrabAttemptResult result = region.attemptToGrabNeighbouringMemory(
+			    stealable, spaceSize, amountToExtend, amountToExtend, thingNotToStealFrom, currentTraversalNo, true);
+
+			// We also told that function to steal the initial main Stealable we are looking at, once it has ascertained that there is enough memory in total.
+			// Previously I attempted to have it steal everything but that central Stealable, and we would steal that afterwards, down below, but this could go wrong
+			// as thefts occurring in the above call to attemptToGrabNeighbouringMemory() could themselves cause other memory to be deallocated or shortened -
+			// and what if this happened to our main, central Stealable before we actually steal it?
+			// This was certainly a problem in automated testing, though I haven't quite wrapped my head around whether this would quite occur under real operation -
+			// but oh well, there is no harm in taking the safe option.
+
+			// If that couldn't be done (in which case the original, central Stealable won't have been stolen either), move on to next Stealable to assess
+			if (!result.address) {
+				if (result.longestRunFound > longestRunSeenInThisQueue) {
+					longestRunSeenInThisQueue = result.longestRunFound;
+				}
+				stealable = static_cast<Stealable*>(reclamation_queue_[q].getNext(stealable));
+				continue;
+			}
+
+			newSpaceAddress = result.address;
```

Comment:
Just to confirm - this is only reachable if we've either stolen enough memory at once or if we've successfully also stolen neighbouring memory?

Reply:
Correct, or we've found memory that can be stolen but has not yet been (the found boolean flag).